src/components/tl/ucp/allgather/allgather_xgvmi.c

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright(c) 2021-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_ucp_ep.h"
+#include "tl_ucp_coll.h"
+#include "tl_ucp_dpu_offload.h"
+
+void ucc_tl_ucp_dpu_xgvmi_rdma_progress_allgather(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t    *task           = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_rank_t            size           = (ucc_rank_t)task->subset.map.ep_num;
+    ucc_datatype_t        dtype          = TASK_ARGS(task).src.info.datatype;
+    size_t                dt_size        = ucc_dt_size(dtype);
+    ucc_count_t           count          = coll_task->bargs.args.src.info.count;
+    ucc_rank_t            host_team_size = size;
+    ucc_base_team_t      *base_team      = coll_task->team;
+    ucc_tl_ucp_team_t    *tl_team        = ucc_derived_of(base_team, ucc_tl_ucp_team_t);
+    ucc_coll_task_t      *allgather_task = task->dpu_xgvmi.allgather_task;
+    ucc_tl_ucp_context_t *tl_ctx         = UCC_TL_UCP_TEAM_CTX(tl_team);
+    ucp_request_param_t   req_param      = {0};
+    int                   i              = 0;
+    ucc_rank_t            rank           = UCC_TL_TEAM_RANK(tl_team);
+    size_t                data_size      = (count * dt_size);
+    ucs_status_ptr_t     *requests       = task->dpu_xgvmi.requests;
+    int                  *posted         = &task->dpu_xgvmi.gets_posted;
+    int                  *completed      = &task->dpu_xgvmi.gets_completed;
+    void                 *src_addr;
+    void                 *dst_addr;
+    ucp_rkey_h            rkey;
+    ucp_ep_h              ep;
+    ucc_rank_t            offset;
+
+    if (allgather_task != NULL) {
+        ucc_tl_ucp_dpu_xgvmi_key_exchange_progress(coll_task);
+        return;
+    }
+
+    req_param.op_attr_mask |= UCP_OP_ATTR_FIELD_MEMH;
+
+    for (i = *posted; i < host_team_size; i++) {
```

Comment:
On the first entry `*posted` will be 0, but after that it will be equal to `host_team_size`. It's possible to make a standalone start function, but I thought it would be less code to reuse `ucc_tl_ucp_dpu_xgvmi_start` for both alltoall and allgather. `ucc_tl_ucp_dpu_xgvmi_start` is 70 lines, and I think for alltoall and allgather to have their own start function to do this suggestion, those 70 lines would have to be duplicated